[GEP-25] Namespaced Cloud Profiles (#9142)
gardener/gardener: docs/proposals/25-namespaced-cloud-profiles.md

status: implementable
authors:
- "@benedictweis"
- "@timuthy"
reviewers:
- "@rfranzke"
- "@ScheererJ"
---

# GEP-25: Namespaced Cloud Profiles

## Table of Contents

- [GEP-25: Namespaced Cloud Profiles](#gep-25-namespaced-cloud-profiles)
  - [Table of Contents](#table-of-contents)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Context](#context)
    - [Current State](#current-state)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
  - [Proposal](#proposal)
    - [Approach](#approach)
    - [Manifest](#manifest)
    - [Rendering](#rendering)
    - [Custom RBAC verb](#custom-rbac-verb)
    - [Adjusting `Shoot`s `cloudProfileName` field](#adjusting-shoots-cloudprofilename-field)
    - [Migration path](#migration-path)
  - [Outlook](#outlook)
    - [Cross-project sharing](#cross-project-sharing)
    - [Multi-Level inheritance](#multi-level-inheritance)
    - [`NamespacedCloudProfile`s replacing regular `CloudProfile`s](#namespacedcloudprofiles-replacing-regular-cloudprofiles)
  - [Alternatives](#alternatives)
    - [Arbitrary Value Fields](#arbitrary-value-fields)
    - [Namespaced Cloud Profiles by Selection](#namespaced-cloud-profiles-by-selection)

## Summary

> [CloudProfiles](https://github.com/gardener/gardener/blob/master/docs/concepts/apiserver.md#cloudprofiles) are non-namespaced objects that are managed centrally by Gardener operators. They usually don't only contain global configuration, but options that are relevant to certain projects only (e.g. special machine types). This increases the operation burden for operators and clutters `CloudProfile` objects. On the other hand, users are blocked until the requested special configuration is rolled out to the desired landscapes.

Reviewer: ```suggestion
[CloudProfiles](https://github.com/gardener/gardener/blob/master/docs/concepts/apiserver.md#cloudprofiles) are non-namespaced objects that are managed centrally by Gardener operators. They usually contain not only global configuration, but options that are relevant to certain projects only (e.g. special machine types). This increases the operation burden for operators and clutters `CloudProfile` objects. On the other hand, users are blocked until the requested special configuration is rolled out to the desired landscapes.
```

This GEP proposes a mechanism that allows project administrators to create `NamespacedCloudProfile`s that are only visible in the project they belong to.

## Motivation

### Context

`CloudProfile`s are an integral component of Gardener for managing shoot clusters. They are currently managed centrally by Gardener operators and can be consumed by any Gardener user. However, some teams require frequent changes to a `CloudProfile`, mainly for the following reasons:

1. Testing with different machine types than the ones present in the `CloudProfile`.
2. Need to use different volume types than the ones present in the `CloudProfile`.

> 3. Extending the expiration date of Kubernetes versions. Given that the `CloudProfile` is a cluster-scoped resource, it is currently not possible to extend the expiration date for shoots in one project but only centrally for all projects.

Reviewer 2: Do you mean extending the expiration date of `Kubernetes` version or the expiration date for `Shoots` ?

Author: Extending the expiration date of Kubernetes versions that are defined in `CloudProfile`s is meant.

4. Extending the expiration date for machine images. For the same reasons as extending the expiration date of the Kubernetes versions.

### Current State

> `CloudProfile`s are non-namespaced resources. This means that in a typical Gardener installation, only a handful of `CloudProfile`s (typically one per cloud provider) exist. They can be consumed by any shoot cluster. Consequently, when a project requires changes to `CloudProfile`s for any of the reasons mentioned in the [context section](#context), they are changed for the entire Gardener landscape. Since some projects might require frequent changes, this becomes quite a cumbersome process on both, the operators' and users' sides.

Reviewer 2: ```suggestion
`CloudProfile`s are non-namespaced resources. This means that in a typical Gardener installation, only a handful of `CloudProfile`s (typically one per cloud provider) exist. They can be consumed by any shoot cluster. Consequently, when a project requires changes to `CloudProfile`s for any of the reasons mentioned in the [context section](#context), they are changed for the entire Gardener landscape. Since some projects might require frequent changes, this becomes quite a cumbersome process on both, the operator and user side.
```

### Goals

- Reduce load on Gardener operators for maintaining `CloudProfile`s
- Possibly reduce the time a team has to wait until a change in a `CloudProfile` is reflected in the landscape.
- Make it so that project-scoped information in the `CloudProfile` is only visible to the relevant project
- Full backward compatibility

### Non-Goals

- Automate the approval process for changes in `CloudProfile`s
- Add different, possibly unsupported Kubernetes versions or machine image versions and names

## Proposal

> It is proposed to implement a solution that enables project administrators to create custom (namespaced) `CloudProfile`s. These `CloudProfile`s would be consumed by project users of the project they were created in.

Reviewer 2: ```suggestion
It is proposed to implement a solution that enables project administrators to create custom (namespaced) `CloudProfile`s. These `CloudProfile`s would be consumed by users of the project they were created in.
```

              expirationDate: 2023-08-8T23:59:59Z
            - version: 15.4
            - version: 14.4
            - version: 13.6
      machineTypes: 
        - name: m5.large
          cpu: "4"
          gpu: "0"
          memory: 8Gi
        - name: m5.xlarge
          cpu: "8"
          gpu: "0"
          memory: 16Gi
      volumeTypes:
        - name: ab6
          class: premium
          usable: true
        - name: gp3
          class: standard
          usable: true
```

The rendering is done by a new custom controller registered to the `gardener-controller-manager`. Merge conflicts can not arise during the merge process as they are caught by static validation and an admission plugin for validating the `NamespacedCloudProfile` object.

### Custom RBAC verb

To prevent users from entering arbitrary values in the `kubernetes` and `machineImages` fields, two custom RBAC verbs may be introduced. It can then be checked if the user that is creating or updating the `kubernetes` or `machineImages` field is authorized to do so.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: namespaced-cloud-profile-modify-special-fields
rules:

>   - apiGroups: [""]
>     resources: ["NamespacedCloudProfile"]

Reviewer 3: ```suggestion
  - apiGroups: ["core.gardener.cloud"]
    resources: ["namespacedcloudprofiles"]
```

An example of the new field can be found here:

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: my-shoot
  namespace: project-xyz
  ...
spec:
  ...
  cloudProfile:
    kind: NamespacedCloudProfile
    name: aws-profile-xyz
  ...
```

### Migration path

`CloudProfile`s are not just a central Gardener resource but are also used by Gardener's provider extensions. Therefore, enabling `NamespacedCloudProfiles` is a multi-step process.

1. Rollout `NamespacedCloudProfile` API and validations, rollout Shoot API changes, add a feature gate for both
2. Adjust all provider extensions to understand both `CloudProfile`s as well as `NamespacedCloudProfiles`
3. Migrate `Shoot`s `cloudProfileName` to the `cloudProfile` field but keep `cloudProfileName` in the API for now
4. Enable `NamespacedCloudProfiles` in Gardener

## Outlook

There are a couple of key features to this GEP that fit well into Gardeners development but are not included immediately to keep the scope in line with the concrete defined use cases. These features might be implemented within future GEPs or PRs and would certainly add value to Gardener.

### Cross-project sharing

> A use case could be defined where a `NamespacedCloudProfile` might want to be shared across multiple projects and not just be used within the project it was created in. Especially when taking a broader view of Gardeners development with Private Seeds and Cloud in Country, this feature is probably going to be necessary at some point.

Reviewer 2: ```suggestion
A use case could be defined where a `NamespacedCloudProfile` might want to be shared across multiple projects and not just within the project it was created in. Especially when taking a broader view of Gardeners development with Private Seeds and Cloud in Country, this feature is probably going to be necessary at some point.
```

  cloudProfile:
    kind: NamespacedCloudProfile
    name: aws-profile-xyz
    namespace: other-project-abc
  ...
```

### Multi-Level inheritance

A `NamespacedCloudProfile` could inherit from a `NamespacedCloudProfile` that already inherits from a `CloudProfile`. This would enable reusing `NamespacedCloudProfile`s and would aid in deduplication within fields of `NamespacedCloudProfiles`. This feature can be implemented easily but is also excluded for now to stick to the defined use cases.

When enabling multi-level inheritance, the `NamespacedCloudProfile`s `parent` field should also be adjusted to allow for a `NamespacedCloudProfile` as a parent.

```yaml
parent:
  kind: <NamespacedCloudProfile | CloudProfile>
  name: <objects name>
```

In combination with [cross-project sharing](#cross-project-sharing), the `parent` field should also allow for a namespace to be defined with the same reasoning as in [cross-project sharing](#cross-project-sharing).

```yaml
parent:
  kind: <NamespacedCloudProfile | CloudProfile>
  name: <objects name>
  namespace: <(optional) objects namespace if outside project namespace>
```

### `NamespacedCloudProfile`s replacing regular `CloudProfile`s

> In the future, once `NamespacedCloudProfiles` have established themselves and found good use in the Gardener landscape and amongst its users, they could replace regular `CloudProfile`s entirely. For this, they would have to include all fields of a `CloudProfile`. Replacing `CloudProfile`s with `NamespacedCloudProfile`s has several benefits. Firstly, Gardener maintainers don't have to maintain, test and operate two almost identical objects. Secondly, it would allow for the central "Gardener provided" `CloudProfile`s to use inheritance. This could enable deduplication in our central `CloudProfiles` as common values do not have to be copied from one `CloudProfile` to another `CloudProfile`. Additionally, it could enable both cloud, as well as on-prem infrastructures to only be visible to defined projects and not to every landscape user.

Reviewer 4: I am not sure whether that would be sensible. Then the project admins would have the burden to find out and maintain their namespaced cloud profiles with all their values for the hyperscalers, no?

I was more thinking that also our cluster cloud profiles support inheritance, i.e. something like:
cluster root cloud profile (with Kubernetes versions for instance) <- cluster hyperscaler cloud profiles (like today) <- namespaced cloud profiles/hierarchy (optional)

Anyway, that can be done later as you said. Thank you.

## Alternatives

### Arbitrary Value Fields

> Instead of specifying a `NamespacedCloudProfile` resource, an end user could be allowed to enter arbitrary names in fields such as `machineTypes`. The entry would then, if the user enters a wrong value, throw an error when provisioning the resource at the cloud provider. However, this approach does not seem feasible as the metadata that is specified in a `CloudProfile` like the number of CPUs and amount of RAM is used in the trial clusters (see [Shoot Qutas](https://github.com/gardener/gardener/blob/master/docs/concepts/apiserver.md#shoot-quotas)) to validate quotas and to enable the scale-from-zero feature. Therefore, an exception would need to be developed for this specific, and possibly other, use cases.

Reviewer: ```suggestion
Instead of specifying a `NamespacedCloudProfile` resource, an end user could be allowed to enter arbitrary names in fields such as `machineTypes`. The entry would then, if the user enters a wrong value, throw an error when provisioning the resource at the cloud provider. However, this approach does not seem feasible as the metadata that is specified in a `CloudProfile` like the number of CPUs and amount of RAM is used in the trial clusters (see [Shoot Quotas](https://github.com/gardener/gardener/blob/master/docs/concepts/apiserver.md#shoot-quotas)) to validate quotas and to enable the scale-from-zero feature. Therefore, an exception would need to be developed for this specific, and possibly other, use cases.
```

    - name: suse-chost
      versions:
        - version: 16.4
          expirationDate: 2023-08-8T23:59:59Z
  machineTypes: 
    - name: m5.xlarge
      cpu: "8"
      gpu: "0"
      memory: 16Gi
  volumeTypes:
    - name: ab6
      class: premium
      usable: true
status:
  cloudProfile:
    apiVersion: core.gardener.cloud/v1beta1
    kind: CloudProfile
    spec:
      type: aws
      kubernetes:
        versions:
          - version: 1.27.1
          - version: 1.26.3
          - version: 1.25.8
          - version: 1.24.6
          - version: 1.28.6
            expirationDate: 2024-06-06T01:02:03Z
      machineImages:
        - name: suse-chost
          versions:
            - version: 16.4
              expirationDate: 2023-08-8T23:59:59Z
            - version: 15.4
            - version: 14.4
            - version: 13.6
      machineTypes: 
        - name: m5.large
          cpu: "4"
          gpu: "0"
          memory: 8Gi
        - name: m5.xlarge
          cpu: "8"
          gpu: "0"
          memory: 16Gi
      volumeTypes:
        - name: ab6
          class: premium
          usable: true
        - name: gp3
          class: standard
          usable: true
```

> The rendering is done by a new custom controller registered to the `gardener-controller-manager`. Merge conflicts can not arise during the merge process as they are caught by static validation and an admission plugin for validating the `NamespacedCloudProfile` object.

Reviewer 5: 🤔 does this mean a `NamespacedCloudProfile` can disallow an update of the parent `CloudProfile` that would introduce a merge conflict?
